rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/model/ParametricFileFacts.java

```diff
@@ -108,66 +134,228 @@ public void close(ISourceLocation loc) {
+                VersionedTree tree, List<Diagnostic> messages) {
+
+            var newer = new VersionedDiagnostics(tree, messages);
+            if (VersionedDiagnostics.setIfNewer(box, newer)) {
+                sendDiagnostics();
+            }
+            return newer;
         }
 
-        public void invalidate(boolean isClosing) {
+        private void invalidate(ParametricSummaryBridge summary, boolean isClosing) {
             summary.invalidate(isClosing);
         }
 
-        public void calculate() {
-            summary.calculateSummary();
-            summary.getMessages().thenAccept(this::reportTypeCheckerMessages);
+        public void invalidateAnalyzer(boolean isClosing) {
+            invalidate(summaryAnalyzer, isClosing);
+        }
+
+        public void invalidateBuilder(boolean isClosing) {
+            invalidate(summaryBuilder, isClosing);
+        }
+
+        /**
+         * @param version the version of the file for which summary calculation
+         * is currently requested
+         * @param latestVersion the last version of the file for which summary
+         * calculation was previously requested
+         * @param delay the duration after which the current request for summary
+         * calculation will be granted, unless another request is made in the
+         * meantime (in which case the current request is abandoned)
+         * @param calculation the actual summary calculation
+         * @return a future that provides optional diagnostics that result from
+         * the summary calculation
+         */
+        private CompletableFuture<Optional<VersionedDiagnostics>> calculate(
+                int version, AtomicInteger latestVersion, Duration delay,
+                Supplier<CompletableFuture<Optional<VersionedDiagnostics>>> calculation) {
+
+            latestVersion.set(version);
+            var delayed = CompletableFuture.delayedExecutor(delay.toMillis(), TimeUnit.MILLISECONDS, exec);
+            return CompletableFuture.supplyAsync(() -> {
+                // If no new call to `calculate` has been made after `delay` has
+                // passed (i.e., `lastVersion` hasn't changed in the meantime),
+                // then run the calculation. Else, abandon this calculation.
+                if (latestVersion.get() == version) {
+                    return calculation.get();
+                } else {
+                    return CompletableFuture.completedFuture(Optional.<VersionedDiagnostics>empty());
+                }
+            }, delayed).thenCompose(Function.identity());
+        }
+
+        public void calculateAnalyzer(int version, Duration delay) {
+            calculateAnalyzerLatestResult = calculate(version, latestVersionCalculateAnalyzer, delay, () -> {
+                var analysis = summaryAnalyzer.calculateSummary();
+                var analysisTree = analysis.input;
+                var analysisMessages = analysis.output.get().thenApply(Supplier::get);
```

Comment:
this is quite a trainwreck, what is happening? can we do something to improve the readability of this?

Reply:
Yeah... The type of `analysis.output` is `InterruptibleFuture<Lazy<List<Diagnostic>>>`, but the type of `analysisMessages` is `CompletableFuture<List<Diagnostic>>`, so the conversion is a bit messy. I added a new helper method to make it easier to digest.

Reply:
it got better with the new unwrap function.